minidump-processor/src/stackwalker/amd64.rs

```diff
@@ -0,0 +1,240 @@
+    // the address pointed to by the called procedure's frame pointer.  The
+    // calling procedure's stack pointer (%esp) is 8 higher than the value
+    // of the called procedure's frame pointer at the time the calling
+    // procedure made the CALL: 4 bytes for the return address pushed by the
+    // CALL itself, and 4 bytes for the callee's PUSH of the caller's frame
+    // pointer.
+    //
+    // %ip_new = *(%bp_old + ptr)
+    // %sp_new = %bp_old + ptr
+    // %bp_new = *(%bp_old)
+
+    let caller_ip = stack_memory.get_memory_at_address(last_bp as u64 + POINTER_WIDTH as u64)?;
+    let caller_bp = stack_memory.get_memory_at_address(last_bp as u64)?;
+    let caller_sp = last_bp + POINTER_WIDTH * 2;
+    let caller_ctx = CONTEXT_AMD64 {
+        rip: caller_ip,
+        rsp: caller_sp,
+        rbp: caller_bp,
+        ..CONTEXT_AMD64::default()
+    };
+    let mut valid = HashSet::new();
+    valid.insert(INSTRUCTION_REGISTER);
+    valid.insert(STACK_POINTER_REGISTER);
+    valid.insert(FRAME_POINTER_REGISTER);
+    let context = MinidumpContext {
+        raw: MinidumpRawContext::Amd64(caller_ctx),
+        valid: MinidumpContextValidity::Some(valid),
+    };
+    let mut frame = StackFrame::from_context(context, FrameTrust::FramePointer);
+    adjust_instruction(&mut frame, caller_ip);
+    Some(frame)
+}
+
+fn get_caller_by_cfi(
+    _ctx: &CONTEXT_AMD64,
+    _valid: &MinidumpContextValidity,
+    _trust: FrameTrust,
+    _stack_memory: &MinidumpMemory,
+    _modules: &MinidumpModuleList,
+) -> Option<StackFrame> {
+    // TODO
+    None
+}
+
+fn get_caller_by_scan(
+    ctx: &CONTEXT_AMD64,
+    valid: &MinidumpContextValidity,
+    trust: FrameTrust,
+    stack_memory: &MinidumpMemory,
+    modules: &MinidumpModuleList,
+) -> Option<StackFrame> {
+    // Stack scanning is just walking from the end of the frame until we encounter
+    // a value on the stack that looks like a pointer into some code (it's an address
+    // in a range covered by one of our modules). If we find such an instruction,
+    // we assume it's an ip value that was pushed by the CALL instruction that created
+    // the current frame. The next frame is then assumed to end just before that
+    // ip value.
+    let last_bp = match valid {
+        MinidumpContextValidity::All => Some(ctx.rbp),
+        MinidumpContextValidity::Some(ref which) => {
+            if !which.contains(STACK_POINTER_REGISTER) {
+                return None;
+            }
+            if which.contains(FRAME_POINTER_REGISTER) {
+                Some(ctx.rbp)
+            } else {
+                None
+            }
+        }
+    };
+    // TODO: pointer-align this..? Does CALL push aligned ip values? Is sp aligned?
+    let last_sp = ctx.rsp;
+
+    let default_scan_range = 40;
+    let extended_scan_range = default_scan_range * 4;
+
+    // Breakpad gives more search range to the first frame of the stack for
+    // whatever reason, so let's duplicate that.
```

Comment:
FWIW, the reasoning [is partially explained here](https://chromium.googlesource.com/breakpad/breakpad/+/374e8dcfa7f53f3af6916b54c26217849545bb2b), and [partially here](https://breakpad.appspot.com/501002/). It boils down to the fact that we encountered a lot of crashes where the instruction pointer was off in the weeds so CFI unwinding didn't work, and the most plausible return address seemed to be way farther down the stack.